src/cli/Cli.ts

```diff
@@ -158,109 +158,107 @@ export class Cli {
-            listener.stderr(message);
-          }
-          logErr.push(message);
+    const log: string[] = [];
+    const logErr: string[] = [];
+    const logger: Logger = {
+      log: (message: any): void => {
+        const formattedMessage = Cli.formatOutput(message, args.options);
+        if (listener && listener.stdout) {
+          listener.stdout(formattedMessage);
         }
-      };
-
-      if (args.options.debug) {
-        const message = `Executing command ${command.name} with options ${JSON.stringify(args)}`;
+        log.push(formattedMessage);
+      },
+      logRaw: (message: any): void => {
+        const formattedMessage = Cli.formatOutput(message, args.options);
+        if (listener && listener.stdout) {
+          listener.stdout(formattedMessage);
+        }
+        log.push(formattedMessage);
+      },
+      logToStderr: (message: any): void => {
         if (listener && listener.stderr) {
           listener.stderr(message);
         }
         logErr.push(message);
       }
+    };
 
-      // store the current command name, if any and set the name to the name of
-      // the command to execute
-      const cli = Cli.getInstance();
-      const parentCommandName: string | undefined = cli.currentCommandName;
-      cli.currentCommandName = command.getCommandName();
-      // store the current logger if any
-      const currentLogger: Logger | undefined = request.logger;
-
-      command.action(logger, args as any, (err: any): void => {
-        // restore the original command name
-        cli.currentCommandName = parentCommandName;
-        // restore the original logger
-        request.logger = currentLogger;
-
-        if (err) {
-          return reject({
-            error: err,
-            stderr: logErr.join(os.EOL)
-          });
-        }
+    if (args.options.debug) {
+      const message = `Executing command ${command.name} with options ${JSON.stringify(args)}`;
+      if (listener && listener.stderr) {
+        listener.stderr(message);
+      }
+      logErr.push(message);
+    }
 
-        resolve({
-          stdout: log.join(os.EOL),
-          stderr: logErr.join(os.EOL)
-        });
-      });
+    // store the current command name, if any and set the name to the name of
+    // the command to execute
+    const cli = Cli.getInstance();
+    const parentCommandName: string | undefined = cli.currentCommandName;
+    cli.currentCommandName = command.getCommandName();
+    // store the current logger if any
+    const currentLogger: Logger | undefined = request.logger;
+
+    try {
+      await command.action(logger, args as any);
+    }
+    catch (err: any) {
+      throw {
+        error: err,
+        stderr: logErr.join(os.EOL)
+      };
+    }
+    finally {
+      // restore the original command name
+      cli.currentCommandName = parentCommandName;
+      // restore the original logger
+      request.logger = currentLogger;
+    }
+
+    return ({
```

Comment:
This should go in the `try` block so that it only executes when running the command succeeded